### PR TITLE
CLOUDP-95181: version is missing

### PR DESCRIPTION
**Pull request by FlorentinaSimlinger**
This PR adds Version to the Status field and correspondingly changes the tests.

### All Submissions:

* [ ] Have you opened an Issue before filing this PR?
* [ ] Have you signed our [CLA](https://www.mongodb.com/legal/contributor-agreement)?
* [ ] Have you signed all of your commits?
* [ ] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?
* [ ] Put `closes #XXXX` in your comment to auto-close the issue that your PR fixes (if such).

### CODE REVIEW

File: controllers/mongodb_status_options.go [hunk 1]
```diff
@@ -22,7 +22,7 @@ const (
 	None  severity = "NONE"
 )
 
-// optionBuilder is in charge of constructing a slice of options that
+// OptionBuilder is in charge of constructing a slice of options that
```
Comment (irajdeep) [inline controllers/mongodb_status_options.go:25]: this should match the function signature -- so should be `optionBuilder` -- let's revert this change.

File: controllers/mongodb_status_options.go [hunk 2]
```diff
@@ -33,7 +33,7 @@ func (o *optionBuilder) GetOptions() []status.Option {
 	return o.options
 }
 
-// options returns an initialized optionBuilder
+// StatusOptions returns an initialized optionBuilder
```
Comment (irajdeep) [inline controllers/mongodb_status_options.go:36]:
```suggestion
// statusOptions returns an initialized optionBuilder
```

File: controllers/mongodb_status_options_test.go [hunk 3]
```diff
@@ -23,30 +23,42 @@ func TestMongoUriOption_ApplyOption(t *testing.T) {
 }
 
 func TestOptionBuilder_RunningPhase(t *testing.T) {
-	mdb := newReplicaSet(3, "my-rs", "my-ns")
+	mdb := newReplicaSet(3, "4.2.6", "my-rs", "my-ns")
```
Comment (irajdeep) [inline controllers/mongodb_status_options_test.go:26]: Can we make `4.2.6` a constant? -- would be easier to make changes later

File: controllers/mongodb_status_options_test.go [hunk 4]
```diff
@@ -23,30 +23,42 @@ func TestMongoUriOption_ApplyOption(t *testing.T) {
...
 }
 
-func newReplicaSet(members int, name, namespace string) mdbv1.MongoDBCommunity {
+func TestVersion_ApplyOption(t *testing.T) {
+	mdb := newReplicaSet(3, "4.2.6", "my-rs", "my-ns")
+
+	opt := versionOption{
+		version: "4.2.6",
+	}
+
```
Comment (irajdeep) [inline controllers/mongodb_status_options_test.go:55]: nit: let's remove empty line here

File: test/e2e/replica_set_scale/replica_set_scaling_test.go [hunk 5]
```diff
@@ -53,6 +53,7 @@ func TestReplicaSetScaleUp(t *testing.T) {
 			mdbv1.MongoDBCommunityStatus{
 				MongoURI:                   mdb.MongoURI(),
 				Phase:                      mdbv1.Running,
+				Version:                    mdb.Spec.Version,
```
Comment (irajdeep) [inline test/e2e/replica_set_scale/replica_set_scaling_test.go:56]: we can call `mdb.GetMongoDBVersion` here

**Review by irajdeep — COMMENTED** [review]
LGTM overall,  added some minor comments.

File: test/e2e/replica_set_recovery/replica_set_recovery_test.go [hunk 6]
```diff
@@ -58,6 +58,7 @@ func TestReplicaSetRecovery(t *testing.T) {
 			mdbv1.MongoDBCommunityStatus{
 				MongoURI:                   mdb.MongoURI(),
 				Phase:                      mdbv1.Running,
+				Version:                    mdb.Spec.Version,
```
Comment (irajdeep) [inline test/e2e/replica_set_recovery/replica_set_recovery_test.go:61]: we can call mdb.GetMongoDBVersion here

File: test/e2e/mongodbtests/mongodbtests.go [hunk 7]
```diff
@@ -310,6 +310,7 @@ func BasicFunctionality(mdb *mdbv1.MongoDBCommunity) func(*testing.T) {
 			mdbv1.MongoDBCommunityStatus{
 				MongoURI:                   mdb.MongoURI(),
 				Phase:                      mdbv1.Running,
+				Version:                    mdb.Spec.Version,
```
Comment (irajdeep) [inline test/e2e/mongodbtests/mongodbtests.go:313]: we can use `mdb.GetMongoDBVersion` here as well

File: test/e2e/replica_set_scale_down/replica_set_scale_down_test.go [hunk 8]
```diff
@@ -56,6 +56,7 @@ func TestReplicaSetScaleDown(t *testing.T) {
 			mdbv1.MongoDBCommunityStatus{
 				MongoURI:                   mdb.MongoURI(),
 				Phase:                      mdbv1.Running,
+				Version:                    mdb.Spec.Version,
```
Comment (irajdeep) [inline test/e2e/replica_set_scale_down/replica_set_scale_down_test.go:59]: ditto

File: controllers/mongodb_status_options.go [hunk 9]
```diff
@@ -60,6 +60,27 @@ func (m mongoUriOption) GetResult() (reconcile.Result, error) {
 	return result.OK()
 }
 
+func (o *optionBuilder) withVersion(mdb *mdbv1.MongoDBCommunity) *optionBuilder {
```
Comment (bznein) [inline controllers/mongodb_status_options.go:63]: [q] why are we passing `mdb` instead of simply the version string?

File: controllers/replica_set_controller.go [hunk 10]
```diff
@@ -224,7 +224,8 @@ func (r ReplicaSetReconciler) Reconcile(ctx context.Context, request reconcile.R
 			withMongoDBMembers(mdb.AutomationConfigMembersThisReconciliation()).
 			withStatefulSetReplicas(mdb.StatefulSetReplicasThisReconciliation()).
 			withMessage(None, "").
-			withRunningPhase(),
+			withRunningPhase().
+			withVersion(&mdb),
```
Comment (bznein) [inline controllers/replica_set_controller.go:228]: related to the other comment, I think `withVersion(mdb.GetMongoDBVersion())` would be much clearer

**Review by bznein — COMMENTED** [review]
Looks good! I just have one comment!

**Review by rodrigovalin — COMMENTED** [review]
@FlorentinaSimlinger is it possible to display what we have in `spec.version` instead of adding a new attribute to status -- an attribute that will be identical to what we already have in `spec.`


cc'ing @bznein as the reporter.

File: test/e2e/replica_set_scale_down/replica_set_scale_down_test.go (hunk 8 above)
Comment (FlorentinaSimlinger) [reply test/e2e/replica_set_scale_down/replica_set_scale_down_test.go:59]: good catch, I'll check again to find all other occurrences

File: controllers/mongodb_status_options.go (hunk 9 above)
Comment (FlorentinaSimlinger) [reply controllers/mongodb_status_options.go:63]: good point, I'll change it to passing the string

**Comment by irajdeep** [comment]
> is it possible to display what we have in spec.version instead of adding a new attribute to status -- an attribute that will be identical to what we already have in spec.

Do you mean not to have the `version` field in the CR status altogether? I think this is more to align with what we have in the enterprise where we show the `version` in the CR status as well ( -- @bznein could correct here if wrong)

**Comment by bznein** [comment]
> > is it possible to display what we have in spec.version instead of adding a new attribute to status -- an attribute that will be identical to what we already have in spec.
> 
> Do you mean not to have the `version` field in the CR status altogether? I think this is more to align with what we have in the enterprise where we show the `version` in the CR status as well ( -- @bznein could correct here if wrong)

I said "status" in the ticket after a quick discussion with Cian. I don't have a strong opinion on this, I think both solutions are ok

**Comment by irajdeep** [comment]
I think one advantage of having the version in the status is to view it at "one place" when running `kubectl get mdbc` when there are multiple CRs in the cluster.

**Comment by rodrigovalin** [comment]
We originally added `version` and a few other attributes to the status object of the enterprise resources because we were holding a "state"; when we designed the community operator we decided to use the "status" as an informational resource that won't have repeated data to hold a state for the reconcile process. This data is stored mostly in annotations right now.

Considering this design, I think it is better if we use `spec.version` to display the version in the mongodb resource.

> I think one advantage of having the version in the status is to view it at "one place" when running kubectl get mdbc when there are multiple CRs in the cluster.

We can have it displayed with `kubectl get mdbc` if we add `.jsonPath: .spec.version` to `additionalPrinterColumns`

**Comment by irajdeep** [comment]
> We can have it displayed with kubectl get mdbc if we add .jsonPath: .spec.version to additionalPrinterColumns

 SGTM

**Comment by bznein** [comment]
> We can have it displayed with `kubectl get mdbc` if we add `.jsonPath: .spec.version` to `additionalPrinterColumns`

SGTM Although this might be slightly not in sync with the correct status. We update the status when the reconciliation is finished, while `spec.version` would reflect the desired version rather than the current one

**Review by rodrigovalin — APPROVED** [review]
> SGTM Although this might be slightly not in sync with the correct status. We update the status when the reconciliation is finished, while spec.version would reflect the desired version rather than the current one

Ok I think this is the right answer to base our conclussion ... If this is the case, I think we should keep displaying the status field, beause it is the "current" version, while the `spec` one is the "desired" one.

In that case we should be good to merge.

LGTM

**Review by chatton — COMMENTED** [review]
I think displaying the version that will be seen in the describe output is a strong reason to keep it. Maintaining multiple resources could become more difficult to manage without it. It is also quite common to list versions with different resources when relevant. (e.g. nodes etc.)

**Comment by FlorentinaSimlinger** [comment]
> SGTM Although this might be slightly not in sync with the correct status. We update the status when the reconciliation is finished, while spec.version would reflect the desired version rather than the current one

Good point, we could use "lastSuccessfulMondoDBVersion" in annotations then -- that would mean that it accurately shows the version even though we might change it during the reconciliation

**Comment by bznein** [comment]
I just noticed that we didn't include this in the Release Notes, I think it's a small fix but worth mentioning, would you mind doing a quick PR to add it?
